### PR TITLE
Feature/backend coding guidelines

### PR DESCRIPTION
**Pull request by ehrdi**

### CODE REVIEW

**Review by dasheck0 — CHANGES_REQUESTED** [review]

File: docs/backend/basic-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    - **500 Internal Server Error:** Server encountered an error.
+    - ...
+  - **Error Messages:** Provide clear and meaningful error messages.
+  - **Data Format:** Use a consistent format for data, typically JSON.
+- **Explanation:** Clear and consistent responses help clients handle different scenarios effectively.
+
+### 5. **Versioning**
+
+- **Principle:** Use versioning to manage changes in the API without disrupting existing clients.
+- **Approaches:**
```
[the remark this answers is not in the document]
Comment (ehrdi) [reply docs/backend/basic-guidelines.md:176]: we dont. whatever fits the specific needs for the project. let's not constrict the developers too much here

File: docs/backend/basic-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    - **200 OK:** Successful GET, PUT, DELETE request.
+    - **201 Created:** Successful POST request.
+    - **204 No Content:** Successful request that does not return any content.
+    - **400 Bad Request:** Invalid request parameters.
+    - **401 Unauthorized:** Authentication required or failed.
+    - **403 Forbidden:** Client authenticated but not authorized.
+    - **404 Not Found:** Resource not found.
+    - **500 Internal Server Error:** Server encountered an error.
+    - ...
+  - **Error Messages:** Provide clear and meaningful error messages.
```
[the remark this answers is not in the document]
Comment (ehrdi) [reply docs/backend/basic-guidelines.md:169]: I wouldn't make it mandatory, but I can add the reference to it anyway

File: docs/backend/basic-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+- **Principle:** Use versioning to manage changes in the API without disrupting existing clients.
+- **Approaches:**
+  - **URI Versioning:** Include the version in the URL (e.g., `/v1/user`).
+  - **Header Versioning:** Use custom headers to specify the version (e.g., `X-API-Version: 1`).
+- **Explanation:** Versioning ensures backward compatibility and allows for continuous improvement of the API.
+
+### 6. **Pagination, Filtering and Sorting**
+
+- **Principle:** Provide mechanisms to handle large sets of data.
```
[the remark this answers is not in the document]
Comment (ehrdi) [reply docs/backend/basic-guidelines.md:183]: Isn't that a bit too much for this document?

File: docs/backend/basic-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+   - **Principle:** Define the API using a standard specification like OpenAPI/Swagger.
+   - **Explanation:** This helps in generating interactive API documentation and client libraries.
+
+2. **Usage Examples**
+
+   - **Principle:** Include examples of requests and responses for each endpoint.
+   - **Explanation:** Examples make it easier for developers to understand how to interact with the API.
+
+3. **Error Codes**
```
[the remark this answers is not in the document]
Comment (ehrdi) [reply docs/backend/basic-guidelines.md:226]: I don't think we should rely on that

File: docs/backend/basic-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+In addition to API documentation, comprehensive backend documentation covers various aspects of the backend development process.
+
+1. **Architecture Overview**
+
+   - **Principle:** Provide an overview of the backend architecture, including high-level components and their interactions.
+   - **Explanation:** Understanding the architecture helps developers navigate the codebase and make informed design decisions.
+
+2. **Database Schema**
+
+   - **Principle:** Document the structure of the database, including tables, columns, relationships, and constraints.
```
[the remark this answers is not in the document]
Comment (ehrdi) [reply docs/backend/basic-guidelines.md:246]: or you can use dbeaver to generate the ERD

File: docs/backend/basic-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+### Unit Testing:
+
+- **Principle:** Test individual components of the API to ensure they work as expected.
+- **Explanation:** Unit tests help catch bugs early in the development process.
+- **Scope:** Individual functions or methods.
+- **Tools:** [Jest](https://github.com/jestjs/jest) (with [NestJS](https://github.com/nestjs/nest)).
+- **What to Test:**
+  - Correctness of individual functions.
+  - Edge cases and error handling.
```
[the remark this answers is not in the document]
Comment (ehrdi) [reply docs/backend/basic-guidelines.md:294]: do you think that is really necessary?

**Review by dasheck0 — APPROVED** [review]
